SDK/simVis/Projector.cpp

```diff
@@ -132,30 +132,143 @@ static const float DEFAULT_ALPHA_VALUE = 0.1f;
    * will be replaced with osg::Callback, after which these two can be
    * unified.
    */
-  class UpdateProjMatrix : public osg::NodeCallback
+#define SIM_MAX_NODE_PROJECTORS 4
+
+  class ProjectOnNodeUpdater : public osg::NodeCallback
   {
+    using ProjectorObserver = osg::observer_ptr<simVis::ProjectorNode>;
+    using ProjectorObservers = std::vector<ProjectorObserver>;
+
   public:
-    explicit UpdateProjMatrix(simVis::ProjectorNode* node) : proj_(node)
+    ProjectOnNodeUpdater()
     {
       //nop
     }
 
-    void operator()(osg::Node* node, osg::NodeVisitor* nv)
+    //! Adds a projector from this callback. 
+    //! Returns the number of projectors being managed by the callback, or
+    //! -1 on error.
+    int add(simVis::ProjectorNode* node)
+    {
+      OE_SOFT_ASSERT_AND_RETURN(node != nullptr, -1);
+      OE_SOFT_ASSERT_AND_RETURN(projectors_.size() < SIM_MAX_NODE_PROJECTORS, -1);
+
+      prune();
+
+      if (std::find(projectors_.begin(), projectors_.end(), node) == projectors_.end())
+      {
+        projectors_.push_back(node);
+        return projectors_.size();
+      }
+      else
+      {
+        // already installed - do nothing
+        return -1;
+      }
+    }
+
+    //! Removes a projector from this callback
+    //! Return the number of remaining projectors
+    int remove(simVis::ProjectorNode* node)
     {
-      if (proj_ == nullptr)
+      auto iter = std::find(projectors_.begin(), projectors_.end(), node);
+      if (iter != projectors_.end())
+        projectors_.erase(iter);
+      prune();
+      return projectors_.size();
+    }
+
+    //! prune any orphaned nodes from the set
+    void prune()
+    {
+      if (projectors_.empty())
         return;
 
-      osgUtil::CullVisitor* cv = dynamic_cast<osgUtil::CullVisitor*>(nv);
-      osg::ref_ptr<osg::StateSet> ss = new osg::StateSet();
-      osg::Matrixf projMat = cv->getCurrentCamera()->getInverseViewMatrix() * proj_->getTexGenMatrix();
-      ss->addUniform(new osg::Uniform("simProjTexGenMat", projMat));
-      cv->pushStateSet(ss.get());
-      traverse(node, nv);
-      cv->popStateSet();
+      ProjectorObservers keep;
+      for (auto& projector : projectors_)
+      {
+        if (projector.valid())
+          keep.push_back(projector);
+      }
+      projectors_.swap(std::move(keep));
```

Comment:
I am getting a compile error with this line on MSVC 2019. I replaced locally with `projectors_ = keep;`